page_iterator.go

```diff
@@ -0,0 +1,195 @@
+	}
+}
+
+// Iterate traverses all pages and enumerates all items in the current page.
+//
+// Iterate receives a callback function which is called with each item in the current page as an argument. The callback function
+// returns a boolean. To traverse and enumerate all pages always return true and to pause traversal and enumeration
+// return false from the callback.
+//
+// Example
+//      pageIterator := NewPageIterator(resp, reqAdapter, parsableCons)
+//      callbackFunc := func (pageItem interface{}) bool {
+//          fmt.Println(pageitem.GetDisplayName())
+//          return true
+//      }
+//      pageIterator.Iterate(callbackFunc)
+func (pI *PageIterator) Iterate(callback func(pageItem interface{}) bool) {
+	for pI.currentPage != nil {
+		keepIterating := pI.enumerate(callback)
+
+		if !keepIterating {
+			// Callback returned false, stop iterating through pages.
+			return
+		}
+
+		pI.next()
+		pI.pauseIndex = 0 // when moving to the next page reset pauseIndex
+	}
+}
+
+// SetHeaders provides headers for requests made to get subsequent pages
+//
+// Headers in the initial request -- request to get the first page -- are not included in subsequent page requests.
+func (pI *PageIterator) SetHeaders(headers map[string]string) {
+	pI.headers = headers
+}
+
+func (pI *PageIterator) hasNext() bool {
+	if pI.currentPage == nil || pI.currentPage.getNextLink() == nil {
+		return false
+	}
+	return true
+}
+
+func (pI *PageIterator) next() Page {
+	nextPage := pI.getNextPage()
+
+	pI.currentPage = nextPage
+	return nextPage
+}
+
+func (pI *PageIterator) getNextPage() *PageResult {
+	if pI.currentPage.getNextLink() == nil {
+		return nil
+	}
+
+	nextLink, err := url.Parse(*pI.currentPage.getNextLink())
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	requestInfo := abstractions.NewRequestInformation()
+	requestInfo.Method = abstractions.GET
+	requestInfo.SetUri(*nextLink)
+	requestInfo.Headers = pI.headers
+
+	res, err := pI.reqAdapter.SendAsync(*requestInfo, pI.constructorFunc, nil)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	return convertToPage(res)
+}
+
+func (pI *PageIterator) enumerate(callback func(item interface{}) bool) bool {
+	keepIterating := true
+
+	if pI.currentPage == nil {
+		return false
+	}
+
+	pageItems := pI.currentPage.getValue()
+	if pageItems == nil {
+		return false
+	}
+
+	if pI.pauseIndex >= len(pageItems) {
+		return false
+	}
+
+	// start/continue enumerating page items from  pauseIndex.
+	// this makes it possible to resume iteration from where we paused iteration.
+	for i := pI.pauseIndex; i < len(pageItems); i++ {
+		keepIterating = callback(pageItems[i])
+
+		if !keepIterating {
+			// Callback returned false, pause! stop enumerating page items. Set pauseIndex so that we know
+			// where to resume from.
+			// Resumes from the next item
+			pI.pauseIndex = i + 1
+			break
+		}
+	}
+
+	return keepIterating
+}
+
+func convertToPage(response interface{}) *PageResult {
+	ref := reflect.ValueOf(response).Elem()
```

Comment:
can't you use a type assertion instead of reflection?

Reply:
We can't use type assertion here because we don't know the type that will be passed.

Reply:
but the passed type will implement the Page interface, which gives us access to what we need no?

Reply:
No the passed type won't implement the Page interface. The tricky bit is specifying `GetValue` return type. While Page specifies `GetValue() []interface{}` The `GetValue` in the passed object returns a slice of graph specific types and for example `[]interface{}` != `[]graph.User`

Reply:
thanks, resolving